File: packages/go/mattersdk/mode.go
package mattersdk

// Delegated mode (runtime spec >= 322): a member-tied API key holds a
// ProxyType::Scoped(ScopeSet) proxy on its member's account and acts through
// proxy.proxy(member, None, call). Its own account has no balance, so a directly
// signed call is refused with "Inability to pay some fees"; the mode is resolved
// at connect so calls are wrapped.

import (
	"bytes"
	"encoding/hex"
	"errors"
	"fmt"
	"log/slog"
	"os"
	"strings"

	"github.com/centrifuge/go-substrate-rpc-client/v4/scale"
	"github.com/centrifuge/go-substrate-rpc-client/v4/types"
	subkey "github.com/vedhavyas/go-subkey/v2"
)

// pallet-proxy's dispatch surface, as a delegated key uses it.
const (
	proxyPallet        = "Proxy"
	proxyCall          = "Proxy.proxy"
	proxyExecutedEvent = "ProxyExecuted"
	// notProxyError: no proxy for this (key, member) pair, i.e. a revoked or
	// rebound key.
	notProxyError = "NotProxy"
)

// agentKeyStateCall maps a key to its principal.
const agentKeyStateCall = "BudgetsApi_agent_key"

// agentKeyCall detects scoped-key support (spec >= 322). V14 metadata lists calls
// but not runtime APIs, so this call, which shipped with BudgetsApi_agent_key, is
// the locally visible signal.
const agentKeyCall = "Budgets.authorize_agent_key"

// principalEnv names the principal a key acts for when the chain's pointer is stale.
const principalEnv = "MATTER_PRINCIPAL"

// agentKeyLookup lets resolveDelegation be tested without a chain.
type agentKeyLookup interface {
	AgentKey(key []byte) (*Delegation, error)
}

// Delegation is who a client's key acts for, and what it may do.
type Delegation struct {
	// Principal is the member every call actually runs as, and who pays for it.
	Principal []byte
	// Scopes is what the chain says this key may do.
	Scopes ScopeSet
}

// forceProxyTypeNone encodes Option<ProxyType>::None; GSRPC has no generic
// Option[T].
type forceProxyTypeNone struct{}

func (forceProxyTypeNone) Encode(encoder scale.Encoder) error {
	return encoder.PushByte(0)
}

// AgentKey reports who `key` acts for and what it may do, per the chain. A nil
// Delegation means no scoped proxy: the runtime predates scoped keys or the key
// is not registered.
func (c *ChainClient) AgentKey(key []byte) (*Delegation, error) {
	if !supportsAgentKeys(c.meta) {
		return nil, nil
	}
	raw, err := c.stateCall(agentKeyStateCall, key)
	if err != nil {
		// Not absence: returning nil would resolve the client to direct mode and
		// make every later write an unexplained fee refusal.
		return nil, fmt.Errorf("%s: %w", agentKeyStateCall, err)
	}
	return decodeAgentKey(raw)
}

// supportsAgentKeys reports whether this runtime has scoped API keys, via the
// same metadata lookup types.NewCall uses, so gate and encoder cannot disagree.
func supportsAgentKeys(meta *types.Metadata) bool {
	if meta == nil {
		return false
	}
	_, err := meta.FindCallIndex(agentKeyCall)
	return err == nil
}

// decodeAgentKey reads Option<(AccountId32, ScopeSet)>: 37 bytes when Some.
func decodeAgentKey(raw []byte) (*Delegation, error) {
	decoder := scale.NewDecoder(strings.NewReader(string(raw)))
	tag, err := decoder.ReadOneByte()
	if err != nil {
		return nil, fmt.Errorf("%s: reading the Option tag: %w", agentKeyStateCall, err)
	}
	if tag == 0 {
		return nil, nil
	}

	// [32]byte, not types.AccountID, so this does not depend on the metadata's
	// shape for AccountId32.
	var principal [32]byte
	if err := decoder.Decode(&principal); err != nil {
		return nil, fmt.Errorf("%s: decoding the principal: %w", agentKeyStateCall, err)
	}
	var bits types.U32
	if err := decoder.Decode(&bits); err != nil {
		return nil, fmt.Errorf("%s: decoding the scope set: %w", agentKeyStateCall, err)
	}
	return &Delegation{Principal: principal[:], Scopes: ScopeSetFromBits(uint32(bits))}, nil
}

// resolveDelegation asks the chain who `account` acts for.
//
// MATTER_PRINCIPAL overrides the chain for a key whose pointer is stale while its
// proxy stands. Scopes are then unknown, so the full set is assumed: local scope
// checks are off, the runtime alone enforces, and a warning is logged.
func resolveDelegation(chain agentKeyLookup, account []byte, ss58Prefix uint16, log *slog.Logger) (*Delegation, error) {
	if override := strings.TrimSpace(os.Getenv(principalEnv)); override != "" {
		principal, err := decodeAccountText(override)
		if err != nil {
			return nil, err
		}
		log.Warn(
			"acting for an overridden principal without asking the chain; "+
				"local scope checking is disabled and the runtime alone enforces",
			"env", principalEnv, "principal", override)
		return &Delegation{Principal: principal, Scopes: AllScopeSet()}, nil
	}

	delegation, err := chain.AgentKey(account)
	if err != nil || delegation == nil {
		return nil, err
	}
	// Logged once: the first thing to check when debugging a fee rejection.
	log.Info("acting for a member",
		"principal", subkey.SS58Encode(delegation.Principal, ss58Prefix),
		"scopes", delegation.Scopes.String())
	return delegation, nil
}

// decodeAccountText parses MATTER_PRINCIPAL: 0x-hex or SS58.
func decodeAccountText(text string) ([]byte, error) {
	bad := fmt.Errorf(
		"%s is neither 0x-prefixed hex nor a valid SS58 address", principalEnv)
	if strings.HasPrefix(text, "0x") {
		raw, err := hex.DecodeString(text[2:])
		if err != nil || len(raw) != 32 {
			return nil, bad
		}
		return raw, nil
	}
	_, raw, err := subkey.SS58Decode(text)
	if err != nil || len(raw) != 32 {
		return nil, bad
	}
	return raw, nil
}

// callNames resolves an encoded call's index bytes to (pallet, call) names.
func callNames(meta *types.Metadata, index types.CallIndex) (string, string, bool) {
	for _, pallet := range meta.AsMetadataV14.Pallets {
		if !pallet.HasCalls || uint8(pallet.Index) != index.SectionIndex {
			continue
		}
		callType, ok := meta.AsMetadataV14.EfficientLookup[pallet.Calls.Type.Int64()]
		if !ok {
			return "", "", false
		}
		for _, variant := range callType.Def.Variant.Variants {
			if uint8(variant.Index) == index.MethodIndex {
				return string(pallet.Name), string(variant.Name), true
			}
		}
	}
	return "", "", false
}

// refreshOutcome is what a re-read of the chain's grant says about the one this
// client resolved at connect.
type refreshOutcome int

const (
	// refreshUnchanged: the same principal, the same scopes.
	refreshUnchanged refreshOutcome = iota
	// refreshRescoped: still ours, but the scopes moved.
	refreshRescoped
	// refreshGone: revoked, or rebound to a different member.
	refreshGone
)

// afterRefresh compares a fresh grant against the one held. A gone grant stays
// delegated: downgrading to direct would make the next write fail as an
// unexplained fee error.
func afterRefresh(previous, fresh *Delegation) refreshOutcome {
	if previous == nil {
		return refreshUnchanged
	}
	if fresh == nil || !bytes.Equal(fresh.Principal, previous.Principal) {
		return refreshGone
	}
	if fresh.Scopes.Bits() != previous.Scopes.Bits() {
		return refreshRescoped
	}
	return refreshUnchanged
}

// isPoolRejection reports whether the node refused a transaction at validation
// rather than the runtime refusing it at dispatch.
//
// A balance-less key the runtime will not admit is refused here, before any
// block, so a revoked key reports "cannot pay fees" rather than Proxy.NotProxy.
// The JSON-RPC code is authoritative; the text is a fallback.
func isPoolRejection(err error) bool {
	if err == nil {
		return false
	}
	var coded interface{ ErrorCode() int }
	if errors.As(err, &coded) && coded.ErrorCode() == poolRejectionCode {
		return true
	}
	text := err.Error()
	for _, marker := range poolRejectionMarkers {
		if strings.Contains(text, marker) {
			return true
		}
	}
	return false
}

// poolRejectionCode is substrate's "Invalid Transaction" JSON-RPC error.
const poolRejectionCode = 1010

// poolRejectionMarkers match the Rust client's, for when the code is lost.
var poolRejectionMarkers = []string{"1010", "Inability to pay", "InvalidTransaction"}

File: crates/matter-sdk/src/chain/mode.rs
//! Whether this client signs for itself or acts for a member.
//!
//! From runtime spec 322, a member-tied API key holds a
//! `ProxyType::Scoped(ScopeSet)` proxy on the minting member's account and acts
//! only via `proxy.proxy(member, None, call)`. Its own account has no balance,
//! so a direct call is refused with `Inability to pay some fees`. A key with no
//! grant is [`Mode::Direct`].

use matter_sdk_key::{AccountId, ScopeSet};
use subxt::dynamic::Value;
use subxt::utils::AccountId32;
use subxt::{OnlineClient, PolkadotConfig};

use crate::error::{Result, SdkError};

/// Runtime API mapping a key to its principal; absent before spec 322.
const AGENT_KEY_TRAIT: &str = "BudgetsApi";
const AGENT_KEY_METHOD: &str = "agent_key";

/// Names the principal a key acts for when the chain's own pointer is stale.
const PRINCIPAL_ENV: &str = "MATTER_PRINCIPAL";

/// Who this client's signature speaks for.
#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub enum Mode {
    /// The signer acts as itself: a human seed, an HSM key, a legacy
    /// project-tied key, or any key on a pre-322 chain.
    Direct,
    /// The signer is a member-tied API key acting for `principal`, and may make
    /// only the calls `scopes` covers.
    Delegated {
        /// The member every call runs as, and who pays for it.
        principal: AccountId,
        /// What the chain says this key may do.
        scopes: ScopeSet,
    },
}

impl Mode {
    /// The member this client acts for, if any.
    pub fn principal(&self) -> Option<AccountId> {
        match self {
            Mode::Direct => None,
            Mode::Delegated { principal, .. } => Some(*principal),
        }
    }

    /// What this client's key is permitted to do, if the chain knows.
    pub fn scopes(&self) -> Option<ScopeSet> {
        match self {
            Mode::Direct => None,
            Mode::Delegated { scopes, .. } => Some(*scopes),
        }
    }

    /// Whether writes must be wrapped in `proxy.proxy`.
    pub fn is_delegated(&self) -> bool {
        matches!(self, Mode::Delegated { .. })
    }
}

/// Ask the chain who `account` acts for.
///
/// 1. `MATTER_PRINCIPAL`, if set, for a key whose pointer is gone while its
///    proxy stands. Assumes [`ScopeSet::ALL`]: local pre-flight is off and the
///    runtime filter alone enforces.
/// 2. Otherwise [`lookup`]; no grant means [`Mode::Direct`].
pub(super) async fn resolve(
    api: &OnlineClient<PolkadotConfig>,
    account: AccountId,
) -> Result<Mode> {
    if let Some(principal) = principal_override()? {
        tracing::warn!(
            %principal,
            env = PRINCIPAL_ENV,
            "acting for an overridden principal without asking the chain; local scope \
             checking is disabled and the runtime alone enforces"
        );
        return Ok(decide(None, Some(principal)));
    }

    Ok(decide(lookup(api, account).await?, None))
}

/// Pure core of [`resolve`]: an override wins; no grant means [`Mode::Direct`].
pub(super) fn decide(
    lookup: Option<(AccountId, ScopeSet)>,
    override_principal: Option<AccountId>,
) -> Mode {
    if let Some(principal) = override_principal {
        return Mode::Delegated {
            principal,
            scopes: ScopeSet::ALL,
        };
    }
    match lookup {
        Some((principal, scopes)) => Mode::Delegated { principal, scopes },
        None => Mode::Direct,
    }
}

/// Whether this runtime has scoped API keys, read from metadata rather than
/// inferred from a failed call.
pub(super) fn supports_agent_keys(metadata: &subxt::Metadata) -> bool {
    has_runtime_api(metadata, AGENT_KEY_TRAIT, AGENT_KEY_METHOD)
}

fn has_runtime_api(metadata: &subxt::Metadata, trait_name: &str, method: &str) -> bool {
    metadata
        .runtime_api_trait_by_name(trait_name)
        .and_then(|t| t.method_by_name(method))
        .is_some()
}

/// `BudgetsApi_agent_key(key)`: who `key` acts for and what it may do. `None`
/// if the runtime lacks the API or the key is not registered.
pub(super) async fn lookup(
    api: &OnlineClient<PolkadotConfig>,
    key: AccountId,
) -> Result<Option<(AccountId, ScopeSet)>> {
    if !supports_agent_keys(&api.metadata()) {
        return Ok(None);
    }

    let payload = subxt::dynamic::runtime_api_call(
        AGENT_KEY_TRAIT,
        AGENT_KEY_METHOD,
        vec![Value::from_bytes(key.as_bytes())],
    );
    let thunk = api
        .runtime_api()
        .at_latest()
        .await
        .map_err(runtime_api_error)?
        .call(payload)
        .await
        .map_err(runtime_api_error)?;

    // Decode typed rather than walking the `Value` tree by hand.
    let resolved: Option<(AccountId32, u32)> = thunk.as_type().map_err(|e| SdkError::Chain {
        target: format!("{AGENT_KEY_TRAIT}_{AGENT_KEY_METHOD}"),
        detail: format!("could not decode Option<(AccountId, ScopeSet)>: {e}"),
    })?;

    Ok(resolved.map(|(principal, bits)| (AccountId::from(principal.0), ScopeSet::from_bits(bits))))
}

/// What a re-read of the chain's grant says about the one a client holds.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(super) enum Refresh {
    /// The same principal with the same scopes.
    Unchanged,
    /// Still this member, but the scopes moved.
    Rescoped { held: ScopeSet },
    /// Revoked, or rebound to a different member.
    Gone,
}

/// Compare a fresh grant against the one resolved at connect.
///
/// `Gone` must not fall back to [`Mode::Direct`]: the caller would see "cannot
/// pay fees" instead of "key revoked".
pub(super) fn after_refresh(previous: &Mode, fresh: Option<(AccountId, ScopeSet)>) -> Refresh {
    let Mode::Delegated { principal, scopes } = previous else {
        return Refresh::Unchanged;
    };
    match fresh {
        Some((fresh_principal, fresh_scopes)) if fresh_principal == *principal => {
            if fresh_scopes == *scopes {
                Refresh::Unchanged
            } else {
                Refresh::Rescoped { held: fresh_scopes }
            }
        }
        _ => Refresh::Gone,
    }
}

/// Parse `MATTER_PRINCIPAL`, which may be `0x`-hex or SS58.
fn principal_override() -> Result<Option<AccountId>> {
    let raw = match std::env::var(PRINCIPAL_ENV) {
        Ok(value) if !value.trim().is_empty() => value,
        _ => return Ok(None),
    };
    let text = raw.trim();

    let bad = |detail: &str| SdkError::Config {
        detail: format!("{PRINCIPAL_ENV} {detail}"),
    };

    if let Some(hex_body) = text.strip_prefix("0x") {
        let bytes = hex::decode(hex_body).map_err(|_| bad("is not valid hex"))?;
        let bytes: [u8; 32] = bytes
            .try_into()
            .map_err(|_| bad("must be 32 bytes of hex"))?;
        return Ok(Some(AccountId::from(bytes)));
    }

    let account: AccountId32 = text
        .parse()
        .map_err(|_| bad("is neither 0x-prefixed hex nor a valid SS58 address"))?;
    Ok(Some(AccountId::from(account.0)))
}

fn runtime_api_error(e: subxt::Error) -> SdkError {
    SdkError::Chain {
        target: format!("{AGENT_KEY_TRAIT}_{AGENT_KEY_METHOD}"),
        detail: e.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use matter_sdk_key::{Access, Scope};

    use super::*;

    #[test]
    fn direct_mode_exposes_no_principal_and_no_scopes() {
        assert_eq!(Mode::Direct.principal(), None);
        assert_eq!(Mode::Direct.scopes(), None);
        assert!(!Mode::Direct.is_delegated());
    }

    #[test]
    fn delegated_mode_carries_the_principal_and_scopes() {
        let principal = AccountId::from([7u8; 32]);
        let scopes = ScopeSet::single(Scope::Deployments, Access::Write);
        let mode = Mode::Delegated { principal, scopes };
        assert_eq!(mode.principal(), Some(principal));
        assert_eq!(mode.scopes(), Some(scopes));
        assert!(mode.is_delegated());
    }

    #[test]
    fn no_grant_resolves_direct() {
        assert_eq!(decide(None, None), Mode::Direct);
    }

    #[test]
    fn a_grant_resolves_delegated() {
        let principal = AccountId::from([3u8; 32]);
        let scopes = ScopeSet::single(Scope::Secrets, Access::Read);
        assert_eq!(
            decide(Some((principal, scopes)), None),
            Mode::Delegated { principal, scopes }
        );
    }

    #[test]
    fn the_override_wins_and_assumes_every_scope() {
        // The empty set would refuse every call locally.
        let forced = AccountId::from([9u8; 32]);
        let chain_said = (AccountId::from([1u8; 32]), ScopeSet::EMPTY);
        for lookup in [None, Some(chain_said)] {
            assert_eq!(
                decide(lookup, Some(forced)),
                Mode::Delegated {
                    principal: forced,
                    scopes: ScopeSet::ALL,
                }
            );
        }
    }

    #[test]
    fn the_spec330_fixture_declares_the_agent_key_runtime_api() {
        assert!(supports_agent_keys(&super::super::test_metadata()));
    }

    #[test]
    fn an_undeclared_runtime_api_reads_as_pre_322() {
        let metadata = super::super::test_metadata();
        assert!(!has_runtime_api(
            &metadata,
            AGENT_KEY_TRAIT,
            "no_such_method"
        ));
        assert!(!has_runtime_api(&metadata, "NoSuchApi", AGENT_KEY_METHOD));
    }

    #[test]
    fn the_call_and_the_runtime_api_ship_together() {
        // Python and Go cannot see runtime APIs in V14 metadata and gate on
        // Budgets.authorize_agent_key instead.
        let metadata = super::super::test_metadata();
        assert!(supports_agent_keys(&metadata));
        assert!(
            metadata
                .pallet_by_name("Budgets")
                .and_then(|p| p.call_variant_by_name("authorize_agent_key"))
                .is_some(),
            "the call the other bindings gate on is missing from the same metadata"
        );
    }

    #[test]
    fn a_revoked_or_rebound_grant_is_gone_and_a_rescope_reports_the_fresh_set() {
        let principal = AccountId::from([4u8; 32]);
        let narrow = ScopeSet::single(Scope::Deployments, Access::Write);
        let wider = narrow.with(Scope::Secrets, Access::Read);
        let held = Mode::Delegated {
            principal,
            scopes: narrow,
        };

        assert_eq!(after_refresh(&held, None), Refresh::Gone);
        assert_eq!(
            after_refresh(&held, Some((AccountId::from([5u8; 32]), narrow))),
            Refresh::Gone
        );
        assert_eq!(
            after_refresh(&held, Some((principal, wider))),
            Refresh::Rescoped { held: wider }
        );
        assert_eq!(
            after_refresh(&held, Some((principal, narrow))),
            Refresh::Unchanged
        );
    }

    #[test]
    fn a_direct_client_has_no_grant_to_lose() {
        assert_eq!(after_refresh(&Mode::Direct, None), Refresh::Unchanged);
    }
}
